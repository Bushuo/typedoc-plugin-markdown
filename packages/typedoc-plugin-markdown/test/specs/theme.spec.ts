import * as fs from 'fs';
import { TestApp } from '../test-app';

describe(`Theme:`, () => {
  let testApp: TestApp;

  beforeAll(() => {
    testApp = new TestApp(['breadcrumbs.ts', 'theme.ts']);
  });

  describe(`(getUrls)`, () => {
    test(`should getUrls'`, async () => {
      await testApp.bootstrap();
      const urlMappings = testApp.theme.getUrls(testApp.project);
      expect(TestApp.getExpectedUrls(urlMappings)).toMatchSnapshot();
    });

    test(`should getUrls with 'allReflectionsHaveOwnDocument' set`, async () => {
      await testApp.bootstrap({ allReflectionsHaveOwnDocument: true });
      const urlMappings = testApp.theme.getUrls(testApp.project);
      expect(TestApp.getExpectedUrls(urlMappings)).toMatchSnapshot();
    });

    test(`should getUrls with readme 'none'`, async () => {
      await testApp.bootstrap({ readme: 'none' });
      const urlMappings = testApp.theme.getUrls(testApp.project);
      expect(TestApp.getExpectedUrls(urlMappings)).toMatchSnapshot();
    });
  });

  describe(`(filenameSeparator)`, () => {
    test(`should getUrls with custom separator'`, async () => {
      await testApp.bootstrap({ filenameSeparator: '-' });
      const urlMappings = testApp.theme.getUrls(testApp.project);
      expect(TestApp.getExpectedUrls(urlMappings)).toMatchSnapshot();
    });
  });

  describe(`(isOutputDirectory)`, () => {
    let directoryListingSpy: jest.SpyInstance;

    beforeAll(async () => {
      await testApp.bootstrap();
      directoryListingSpy = jest.spyOn(fs, 'readdirSync');
    });

    afterAll(() => {
      directoryListingSpy.mockRestore();
    });

    test(`should test output directory true with all allowed files and directories`, () => {
      directoryListingSpy.mockReturnValue([
        '.DS_Store',
        'README.md',
        'modules.md',
        'classes',
        'enums',
        'interfaces',
        'media',
        'modules',
      ]);
      expect(testApp.theme.isOutputDirectory('/path')).toBeTruthy();
    });

    test(`should test output directory true with some files directories`, () => {
      directoryListingSpy.mockReturnValue([
        'README.md',
        'classes',
        'media',
        'modules',
      ]);
      expect(testApp.theme.isOutputDirectory('/path')).toBeTruthy();
    });

    test(`should test output directory true with just index`, () => {
      directoryListingSpy.mockReturnValue(['README.md']);
      expect(testApp.theme.isOutputDirectory('/path')).toBeTruthy();
    });

    test(`should test output directory false with unkown index`, () => {
      directoryListingSpy.mockReturnValue([
        'Unrecognised.md',
        'classes',
        'enums',
        'interfaces',
        'media',
        'modules',
      ]);
      expect(testApp.theme.isOutputDirectory('/path')).toBeFalsy();
    });

    test(`should test output directory false with hidden files`, () => {
      directoryListingSpy.mockReturnValue([
        '.git',
        'classes',
        'enums',
        'interfaces',
        'media',
        'modules',
      ]);
      expect(testApp.theme.isOutputDirectory('/path')).toBeFalsy();
    });

    test(`should test output directory false with unknown folder`, () => {
      directoryListingSpy.mockReturnValue(['README.md', 'folder']);
      expect(testApp.theme.isOutputDirectory('/path')).toBeFalsy();
    });
  });

  describe(`(directory mappings)`, () => {
    test(`should set default mappings'`, async () => {
      await testApp.bootstrap();
      expect(
        testApp.theme.mappings.map((mapping) => mapping.directory),
      ).toEqual(['classes', 'interfaces', 'enums', 'modules']);
    });

    test(`should set mappings with 'allReflectionsHaveOwnDocument'`, async () => {
      await testApp.bootstrap({ allReflectionsHaveOwnDocument: true });
      expect(
        testApp.theme.mappings.map((mapping) => mapping.directory),
      ).toEqual([
        'classes',
        'interfaces',
        'enums',
        'modules',
        'variables',
        'types',
        'functions',
        'literals',
      ]);
    });
  });
});
