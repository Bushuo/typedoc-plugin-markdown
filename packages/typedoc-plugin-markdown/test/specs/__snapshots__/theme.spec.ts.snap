// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Theme: (entryDocument) should getUrls with custom entryDocument' 1`] = `
Array [
  "modules.md",
  "index.md",
  "modules/breadcrumbs.md",
  "classes/breadcrumbs.breadcrumbs-1.md",
  "classes/breadcrumbs.breadcrumbs-1.md#constructor",
  "classes/breadcrumbs.breadcrumbs-1.md#somemethod",
  "modules/theme.md",
  "modules/theme.md#declarationitema",
  "modules/theme.md#declarationitemb",
  "modules/theme.md#functionitema",
  "modules/theme.md#functionitemb",
  "modules/theme.namespacea.md",
  "classes/theme.namespacea.somenestedclass.md",
  "classes/theme.namespacea.somenestedclass.md#constructor",
  "modules/theme.namespaceb.md",
  "enums/theme.enumitema.md",
  "enums/theme.enumitemb.md",
  "classes/theme.classitema.md",
  "classes/theme.classitema.md#constructor",
  "classes/theme.classitemb.md",
  "classes/theme.classitemb.md#constructor",
  "interfaces/theme.interfaceitema.md",
  "interfaces/theme.interfaceitema.md#prop",
  "interfaces/theme.interfaceitemb.md",
  "interfaces/theme.interfaceitemb.md#prop",
]
`;

exports[`Theme: (filenameSeparator) should getUrls with custom separator' 1`] = `
Array [
  "modules.md",
  "README.md",
  "modules/breadcrumbs.md",
  "classes/breadcrumbs-breadcrumbs-1.md",
  "classes/breadcrumbs-breadcrumbs-1.md#constructor",
  "classes/breadcrumbs-breadcrumbs-1.md#somemethod",
  "modules/theme.md",
  "modules/theme.md#declarationitema",
  "modules/theme.md#declarationitemb",
  "modules/theme.md#functionitema",
  "modules/theme.md#functionitemb",
  "modules/theme-namespacea.md",
  "classes/theme-namespacea-somenestedclass.md",
  "classes/theme-namespacea-somenestedclass.md#constructor",
  "modules/theme-namespaceb.md",
  "enums/theme-enumitema.md",
  "enums/theme-enumitemb.md",
  "classes/theme-classitema.md",
  "classes/theme-classitema.md#constructor",
  "classes/theme-classitemb.md",
  "classes/theme-classitemb.md#constructor",
  "interfaces/theme-interfaceitema.md",
  "interfaces/theme-interfaceitema.md#prop",
  "interfaces/theme-interfaceitemb.md",
  "interfaces/theme-interfaceitemb.md#prop",
]
`;

exports[`Theme: (getUrls) should getUrls with 'allReflectionsHaveOwnDocument' set 1`] = `
Array [
  "modules.md",
  "README.md",
  "modules/breadcrumbs.md",
  "classes/breadcrumbs.breadcrumbs-1.md",
  "classes/breadcrumbs.breadcrumbs-1.md#constructor",
  "classes/breadcrumbs.breadcrumbs-1.md#somemethod",
  "modules/theme.md",
  "modules/theme.namespacea.md",
  "classes/theme.namespacea.somenestedclass.md",
  "classes/theme.namespacea.somenestedclass.md#constructor",
  "modules/theme.namespaceb.md",
  "enums/theme.enumitema.md",
  "enums/theme.enumitemb.md",
  "classes/theme.classitema.md",
  "classes/theme.classitema.md#constructor",
  "classes/theme.classitemb.md",
  "classes/theme.classitemb.md#constructor",
  "interfaces/theme.interfaceitema.md",
  "interfaces/theme.interfaceitema.md#prop",
  "interfaces/theme.interfaceitemb.md",
  "interfaces/theme.interfaceitemb.md#prop",
  "variables/theme.declarationitema.md",
  "variables/theme.declarationitemb.md",
  "functions/theme.functionitema.md",
  "functions/theme.functionitemb.md",
]
`;

exports[`Theme: (getUrls) should getUrls with readme 'none' 1`] = `
Array [
  "README.md",
  "modules/breadcrumbs.md",
  "classes/breadcrumbs.breadcrumbs-1.md",
  "classes/breadcrumbs.breadcrumbs-1.md#constructor",
  "classes/breadcrumbs.breadcrumbs-1.md#somemethod",
  "modules/theme.md",
  "modules/theme.md#declarationitema",
  "modules/theme.md#declarationitemb",
  "modules/theme.md#functionitema",
  "modules/theme.md#functionitemb",
  "modules/theme.namespacea.md",
  "classes/theme.namespacea.somenestedclass.md",
  "classes/theme.namespacea.somenestedclass.md#constructor",
  "modules/theme.namespaceb.md",
  "enums/theme.enumitema.md",
  "enums/theme.enumitemb.md",
  "classes/theme.classitema.md",
  "classes/theme.classitema.md#constructor",
  "classes/theme.classitemb.md",
  "classes/theme.classitemb.md#constructor",
  "interfaces/theme.interfaceitema.md",
  "interfaces/theme.interfaceitema.md#prop",
  "interfaces/theme.interfaceitemb.md",
  "interfaces/theme.interfaceitemb.md#prop",
]
`;

exports[`Theme: (getUrls) should getUrls' 1`] = `
Array [
  "modules.md",
  "README.md",
  "modules/breadcrumbs.md",
  "classes/breadcrumbs.breadcrumbs-1.md",
  "classes/breadcrumbs.breadcrumbs-1.md#constructor",
  "classes/breadcrumbs.breadcrumbs-1.md#somemethod",
  "modules/theme.md",
  "modules/theme.md#declarationitema",
  "modules/theme.md#declarationitemb",
  "modules/theme.md#functionitema",
  "modules/theme.md#functionitemb",
  "modules/theme.namespacea.md",
  "classes/theme.namespacea.somenestedclass.md",
  "classes/theme.namespacea.somenestedclass.md#constructor",
  "modules/theme.namespaceb.md",
  "enums/theme.enumitema.md",
  "enums/theme.enumitemb.md",
  "classes/theme.classitema.md",
  "classes/theme.classitema.md#constructor",
  "classes/theme.classitemb.md",
  "classes/theme.classitemb.md#constructor",
  "interfaces/theme.interfaceitema.md",
  "interfaces/theme.interfaceitema.md#prop",
  "interfaces/theme.interfaceitemb.md",
  "interfaces/theme.interfaceitemb.md#prop",
]
`;

exports[`Theme: (legacy filenames) should getUrls with legacy filenames' 1`] = `
Array [
  "modules.md",
  "README.md",
  "modules/_breadcrumbs_.md",
  "classes/_breadcrumbs_.breadcrumbs-1.md",
  "classes/_breadcrumbs_.breadcrumbs-1.md#constructor",
  "classes/_breadcrumbs_.breadcrumbs-1.md#somemethod",
  "modules/_theme_.md",
  "modules/_theme_.md#declarationitema",
  "modules/_theme_.md#declarationitemb",
  "modules/_theme_.md#functionitema",
  "modules/_theme_.md#functionitemb",
  "modules/_theme_.namespacea.md",
  "classes/_theme_namespacea_.somenestedclass.md",
  "classes/_theme_namespacea_.somenestedclass.md#constructor",
  "modules/_theme_.namespaceb.md",
  "enums/_theme_.enumitema.md",
  "enums/_theme_.enumitemb.md",
  "classes/_theme_.classitema.md",
  "classes/_theme_.classitema.md#constructor",
  "classes/_theme_.classitemb.md",
  "classes/_theme_.classitemb.md#constructor",
  "interfaces/_theme_.interfaceitema.md",
  "interfaces/_theme_.interfaceitema.md#prop",
  "interfaces/_theme_.interfaceitemb.md",
  "interfaces/_theme_.interfaceitemb.md#prop",
]
`;
